MapboxNavigation/NavigationMapView.swift

```diff
@@ -310,15 +312,34 @@ open class NavigationMapView: MGLMapView, UIGestureRecognizerDelegate {
             let duration: TimeInterval = animated ? 1 : 0
             setCamera(newCamera, withDuration: duration, animationTimingFunction: function, edgePadding: padding, completionHandler: nil)
         }
-        
         let duration: TimeInterval = animated ? 1 : 0
-        UIView.animate(withDuration: duration, delay: 0, options: [.curveLinear, .beginFromCurrentState], animations: {
-            self.userCourseView?.center = self.convert(location.coordinate, toPointTo: self)
-        }, completion: nil)
         
-        if let userCourseView = userCourseView as? UserCourseView {
-            userCourseView.update(location: location, pitch: camera.pitch, direction: direction, animated: animated, tracksUserCourse: tracksUserCourse)
+        if let route = routes?.first, let coordinates = route.coordinates, let previousLocation = previousLocation {
+            
+            let points = Polyline(coordinates).sliced(from: previousLocation.coordinate, to: location.coordinate).coordinates.map {
```

Comment:
NavigationMapView’s concept of routes is pretty display-focused. For example, the concept of a “first” route is pretty tied to the idea that it renders on top of the others. By contrast, RouteController is currently wholly responsible for calculating coordinates for the course view. Let’s see if we can push the snapped coordinate generation back down into RouteController and ensure that this method obtains its route from RouteProgress, if it even considers the route at all.

Reply:
I agree with you that things could be moved around here, but a reminder `RouteController` does not do anything UI related. 

We could however look into bisecting lines between points even further in RouteController.

Reply:
> a reminder RouteController does not do anything UI related

That’s true, but RouteController does house the location snapping code that’s primarily used by the user puck. The trick is to make RouteController responsible for anything spatial but make the APIs generic enough that they don’t _have_ to be used with a particular UI feature.